diff --git a/src/platforms/common/jtagtap.c b/src/platforms/common/jtagtap.c
index 3527d68d..e38380f1 100644
--- a/src/platforms/common/jtagtap.c
+++ b/src/platforms/common/jtagtap.c
@@ -40,6 +40,9 @@ static void jtagtap_cycle(bool tms, bool tdi, size_t clock_cycles);
 
 void jtagtap_init(void)
 {
+#ifdef JTAGTAP_PLATFORM_INIT
+        jtagtap_platform_init();
+#endif
 	platform_target_clk_output_enable(true);
 	TMS_SET_MODE();
 
diff --git a/src/platforms/common/swdptap.c b/src/platforms/common/swdptap.c
index 1f171f83..f498e391 100644
--- a/src/platforms/common/swdptap.c
+++ b/src/platforms/common/swdptap.c
@@ -60,6 +60,9 @@ static void swdptap_seq_out_parity(uint32_t tms_states, size_t clock_cycles) __a
 
 void swdptap_init(void)
 {
+#ifdef SWDPTAP_PLATFORM_INIT
+        swdptap_platform_init();
+#endif
 	swd_proc.seq_in = swdptap_seq_in;
 	swd_proc.seq_in_parity = swdptap_seq_in_parity;
 	swd_proc.seq_out = swdptap_seq_out;
