diff --git a/blackmagic/src/rtt.c b/blackmagic/src/rtt.c
index 864953b..65d4e8e 100644
--- a/blackmagic/src/rtt.c
+++ b/blackmagic/src/rtt.c
@@ -129,8 +129,9 @@ static uint32_t fast_search(target_s *const cur_target, const uint32_t ram_start
 	static const uint64_t q = 0x797a9691; /* prime */
 	static const uint32_t stride = 128;
 	uint64_t hash = 0;
-	uint8_t *srch_buf = alloca(hash_len + stride);
+	static uint8_t *srch_buf = NULL;
 
+        if (!srch_buf) srch_buf = alloca(hash_len + stride);
 	memset(srch_buf, 0, hash_len + stride);
 
 	for (uint32_t addr = ram_start; addr < ram_end; addr += stride) {
